thoth/adviser/state.py

```diff
@@ -170,12 +171,93 @@ def get_first_unresolved_dependency(self) -> Tuple[str, str, str]:
                 f"No unresolved dependency found in state: {self!r}"
             ) from exc
 
-    def get_random_unresolved_dependency(self) -> Tuple[str, str, str]:
+    @staticmethod
+    def _termial_function(n: int) -> int:
+        r"""Compute termial function, with red hats off to Donald Knuth.
+
+        Termial function is:
+
+          \sum_{k=0}^n k
+
+        Let's use the solution using binomial coefficient - (n + 1) over 2.
+        """
+        if n < 0:
+            raise ValueError("Termial function used with negative number")
+
+        return ((n**2) + n) // 2
+
+    @staticmethod
+    def _termial_function_solution(x: int) -> int:
+        """Compute solution to termial function.
+
+        Let termial function be:
+
+           (n**2 + n) / 2 = x
+
+        This function computes non-negative ceil(n) so that it satisfies x.
+
+        The solution is optimized to compute a generic parabolic intersection for termial function:
+
+           x1 = (-b + sqrt(b**2 - 4*a*c)) / 2*a
+           x2 = (-b - sqrt(b**2 - 4*a*c)) / 2*a
+
+        Note solution x2 can be discarded based on termial function usage in the sources.
+
+        The original formula:
+
+           a*x**2 + b*x + c = 0
+
+        for solution:
+
+           (n**2)/2 + n/2 = x
+
+        where:
+
+          a = 1/2
+          b = 1/2
+          c = -x
+        """
+        return math.floor(-1/2 + math.sqrt(1/4 + 2*x))
+
+    @classmethod
+    def _random_termial(cls, n: int) -> int:
+        """Compute a random number x such as 0 <= x < n, use termial function not to spread numbers uniformly.
+
+        To prefer lower numbers more often, termial function is used to assign "weights" for
+        numbers - then random uniform is used in conjuction with weights.
+
+        An illustrative example can be cls._random_terminal(4).
+
+          cls._terminal_function(4) == 10
```

Comment:
`_random_termial` and `_termial_function`?

Reply:
`_random_termial` calls `_termial_function` - hence the comment. I will adjust the docstring.

Reply:
are there typos also in the docstring?

Reply:
Agh, right :+1: It's termial, fixed in 354a9a7

Reply:
Then it's ready for me :) Thanks, sorry for being picky :P

Reply:
Actually, I love picky reviews :smile_cat: